xlators/cluster/afr/src/afr-common.c

```diff
@@ -5097,18 +5098,135 @@ afr_lk_unlock(call_frame_t *frame, xlator_t *this)
     return 0;
 }
 
+static gf_boolean_t
+afr_is_need_migrate_lock(afr_local_t *local, unsigned int child_count)
+{
+    unsigned int index = 0;
+    gf_boolean_t is_need_migrate = _gf_false;
+
+    for (index = 0; index < child_count; index++) {
+        if (local->need_migrate_lock[index]) {
+            is_need_migrate = _gf_true;
+            break;
+        }
+    }
+    return is_need_migrate;
+}
+
+static dict_t *
+afr_get_lock_info_from_healthy_brick(afr_fd_ctx_t *fd_ctx, fd_t *fd,
+                                     afr_private_t *priv, afr_local_t *local)
+{
+    dict_t *lockinfo = NULL;
+    void *ptr = NULL;
+    int ret = -1;
+    int index = 0;
+
+    for (index = 0; index < priv->child_count; index++) {
+        if (fd_ctx->opened_on[index] == AFR_FD_OPENED &&
+            priv->child_up[index] && !local->need_migrate_lock[index]) {
+            ret = syncop_fgetxattr(priv->children[index], fd, &lockinfo,
+                                   GF_XATTR_LOCKINFO_KEY, NULL, NULL);
+            if (ret == 0) {
+                break;
+            }
+        }
+    }
+
+    if (ret) {
+        goto out;
+    }
+
+    ret = dict_get_ptr(lockinfo, GF_XATTR_LOCKINFO_KEY, &ptr);
+    if (ptr == NULL) {
+        if (lockinfo != NULL)
+            dict_unref(lockinfo);
+        lockinfo = NULL;
+        goto out;
+    }
+
+out:
+    return lockinfo;
+}
+
+static int
+afr_migrate_locks(afr_private_t *priv, afr_local_t *local)
+{
+    int ret = 0;
+    int index = -1;
+    fd_t *fd = NULL;
+    dict_t *lockinfo = NULL;
+    gf_boolean_t is_migrate_success = _gf_false;
+    afr_fd_ctx_t *fd_ctx = NULL;
+    xlator_t *this = THIS;
+
+    if (local == NULL)
+        return ret;
+
+    fd = local->fd;
+    if (!fd->lk_ctx || !afr_is_need_migrate_lock(local, priv->child_count)) {
+        ret = 0;
+        goto out;
+    }
+
+    fd_ctx = afr_fd_ctx_get(fd, this);
+    if (!fd_ctx) {
+        ret = 0;
+        goto out;
+    }
+
+    lockinfo = afr_get_lock_info_from_healthy_brick(fd_ctx, fd, priv, local);
+    if (lockinfo == NULL) {
+        goto out;
+    }
+
+    local->loc.inode = inode_ref(fd->inode);
+    for (index = 0; index < priv->child_count; index++) {
+        if (!local->need_migrate_lock[index])
+            continue;
+        if (fd_ctx->opened_on[index] == AFR_FD_NOT_OPENED &&
+            priv->child_up[index]) {
```

Comment:
If `opened_on` is `AFR_FD_NOT_OPENED` but the child is not up, we won't enter this `if` branch, but the next `syncop_fsetxattr()` will be executed. It doesn't make sense to try the fsetxattr under these conditions.